Fix consumption of CPU-backed interchange protocol dataframes

shwina opened the pull request:
## Description

Closes #11245. This PR fixes a bug in our code that consumes a protocol DataFrame that is backed by CPU memory.

This enables using the `from_dataframe()` function to construct DataFrames from other libraries:

```python
In [12]: pdf = pd.DataFrame({'x': [1, 2, 3], 'y': [1.0, 2.0, np.nan], 'z': ['a', 'b', 'c']})

In [13]: vdf = vaex.from_dict({'x': [1, 2, 3], 'y': [1.0, 2.0, np.nan]})

In [14]: cudf.from_dataframe(pdf, allow_copy=True)
Out[14]: 
   x    y  z
0  1  1.0  a
1  2  2.0  b
2  3  NaN  c

In [15]: cudf.from_dataframe(vdf, allow_copy=True)
Out[15]: 
   x    y
0  1  1.0
1  2  2.0
2  3  NaN
```



## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

shwina commented:
rerun tests

wence- commented on python/cudf/cudf/tests/test_df_protocol.py:
```diff
@@ -232,3 +238,13 @@ def test_NA_mixed_dtype():
+def test_from_cpu_df(pandas_df):
+    df = pd.DataFrame({"a": [1, 2, 3], "b": ["x", "y", "z"]})
+    with pytest.raises(NotImplementedError):
+        cudf.from_dataframe(df, allow_copy=True)
```
Can we assert this actually worked?

shwina replied:
The goal of this PR was to raise the correct error message when consuming CPU-backed DataFrames. After reading your comment, I figured we might as well implement the bits needed to actually consume them. In the process I discovered a slew of issues that I've tried to address in this PR. I've re-requested your review, but I'll leave comments inline where I think it would be useful.

wence- commented on python/cudf/cudf/core/df_protocol.py:
```diff
@@ -660,13 +660,8 @@ def from_dataframe(
         raise ValueError("`df` does not support __dataframe__")
 
-    return _from_dataframe(df.__dataframe__(allow_copy=allow_copy))
+    df = df.__dataframe__(allow_copy=allow_copy)
```
Although `from_dataframe` previously didn't document the `allow_copy` argument, maybe now is a good time to do so?

bdice replied:
+1, please add parameter/return docstrings to `from_dataframe`.

shwina replied:
Done

wence- commented:
Tests failed in `strings_udf` 
```
tables.pyx:13: in strings_udf._lib.tables.get_character_flags_table_ptr
    ???
E   RuntimeError: Fatal CUDA error encountered at: /workspace/.conda-bld/work/cpp/src/strings/utilities.cu:96: 222 unrecognized error code unrecognized error code
```

I lost track of what was happening there, will that be fixed by a back-merge from trunk?

wence- reviewed (changes requested):
Very minor docstring quibble, otherwise looks good. I suppose the copies happen implicitly when the various `Buffer` objects are created?

shwina commented on python/cudf/cudf/core/df_protocol.py:
```diff
@@ -46,6 +48,14 @@ class _Device(enum.IntEnum):
     ROCM = 10
 
 
+class _MaskKind(enum.IntEnum):
+    NON_NULLABLE = (0,)
+    NAN = (1,)
+    SENTINEL = (2,)
+    BITMASK = (3,)
+    BYTEMASK = 4
```
Prior to this PR, we used raw ints to represent this information.

wence- replied:
Is it deliberate that `BYTEMASK` is an int, but the other mask kinds are tuples? (They both turn into ints on construction, but the tuple instantiation looks kind of weird.

shwina replied:
Fixed to just use ints here.

shwina commented on python/cudf/cudf/core/df_protocol.py:
```diff
@@ -399,32 +408,22 @@ def _get_validity_buffer(
 
         Raises RuntimeError if null representation is not a bit or byte mask.
         """
-
```
It turns out that we were erroneously returning a boolean mask while promising that we were returning a bitmask (the interchange protocol allows returning either; the `col.describe_null` method tells the consumer which one to expect).

I've fixed this method to return a bitmask instead.

shwina commented on python/cudf/cudf/core/df_protocol.py:
```diff
@@ -645,7 +644,13 @@ def __dataframe__(
+    1: _UINTS,
+    2: _FLOATS,
+    20: {8: bool},
+    21: {8: cp.uint8},
```
Strings columns in the protocol are composed of a `uint8` buffer of `data` (chars) and an `int` buffer of `offsets`. The `data` column has a distinct type ID (`21`) that we previously didn't handle.

As a clarifying example, look at the buffers returned by a Pandas dataframe containing a single string column:

```python
In [17]: pd.DataFrame({'a': ['x', 'y', 'z']}).__dataframe__().get_column(0).get_buffers()
Out[17]: 
{'data': (PandasBuffer({'bufsize': 3, 'ptr': 140001666253824, 'device': 'CPU'}),
  (<DtypeKind.STRING: 21>, 8, 'u', '=')),
 'validity': (PandasBuffer({'bufsize': 3, 'ptr': 94467671186912, 'device': 'CPU'}),
  (<DtypeKind.BOOL: 20>, 8, 'b', '=')),
 'offsets': (PandasBuffer({'bufsize': 32, 'ptr': 94467669930400, 'device': 'CPU'}),
  (<DtypeKind.INT: 0>, 64, 'l', '='))}
```

shwina commented on python/cudf/cudf/tests/test_df_protocol.py:
```diff
@@ -7,25 +7,54 @@
+
+
+def assert_validity_equal(protocol_buffer, cudf_buffer, size, null, valid):
+    if null == _MaskKind.BYTEMASK:
```
Comparing boolmasks is easy, just use `assert_eq`.

shwina commented on python/cudf/cudf/tests/test_df_protocol.py:
```diff
@@ -7,25 +7,54 @@
+            as_column(cudf_buffer, dtype="bool"),
+        )
+    elif null == _MaskKind.BITMASK:
+        protocol_mask = _protocol_buffer_to_cudf_buffer(protocol_buffer)
```
Comparing bitmasks is harder because we don't have utilities in Python to compare two bitmasks AFAICT. Instead, we construct two string columns composed entirely of empty strings, assign them the "left" and "right" bitmasks respectively, and then use `assert_eq` to compare the resulting columns.

Open to other suggestions here :-)

bdice replied:
If we construct a column of empty strings, it still requires offsets to be allocated/generated (which are `size_type = int32`). A more lightweight approach would be to use an `int8` data type with this mask.

shwina replied:
Fixed to use an int8 column here instead

bdice reviewed (changes requested):
@shwina This popped back up on my radar after you bumped the branch. If this is still active, I think it only needs a bit more work to be complete?

wence- commented on python/cudf/cudf/core/df_protocol.py:
```diff
@@ -719,49 +725,56 @@ def _protocol_to_cudf_column_numeric(
+    # return it as is.  Otherwise, copy it to the device and return
+    # the resulting buffer.
+    if buf.__dlpack_device__()[0] != _Device.CUDA:
+        if not allow_copy:
+            raise TypeError(
+                "This operation must copy data from CPU to GPU. "
+                "Set `allow_copy=True` to allow it."
+            )
+        else:
+            dbuf = rmm.DeviceBuffer(ptr=buf.ptr, size=buf.bufsize)
+            return _CuDFBuffer(
+                as_buffer(dbuf, exposed=True),
+                protocol_dtype_to_cupy_dtype(data_type),
+                allow_copy,
+            )
```
Style, I (at least) find double-negated conditions hard to parse. Perhaps have the truthy case in the if, and the exceptional case in the else branch:
```
if allow_copy:
    dbuf = ...
    return ...
else:
    raise TypeError(...)
```

shwina replied:
Done

wence- reviewed (commented):
Some minor nitpicks, but otherwise looks good.

vyasr commented:
@shwina what's the status of this PR?

shwina commented:
Thanks @vyasr and apologies all for letting this PR go stale - I've updated the PR to target 23.06, and I believe addressed all the review comments.

@bdice, @wence-, would you be willing to take a final look? Thank you for the suggested improvements!

wence- reviewed (commented):
Implementation looks good, a minor request to add a little more testing in the pandas<->cudf interop tests

shwina commented on python/cudf/cudf/core/df_protocol.py:
```diff
@@ -828,27 +871,32 @@ def _protocol_to_cudf_column_string(
+        rmm.DeviceBuffer(
+            ptr=protocol_buffer.ptr, size=protocol_buffer.bufsize
+        ),
+        exposed=True,
```
Not exactly. The relationship between spilling and copy-on-write is that enabling spilling and copy-on-write simultaneously is disallowed.

wence- replied:
I think what I meant was, in the same way that the buffer is exposed (disallowing spilling) if we take an external reference, it should also be marked as "exposed" such that a copy-on-write of the buffer copies eagerly.

shwina replied:
Ah I see. Yes, I believe marking the buffer `exposed` will also set the `copy_on_write` flag appropriately. @galipremsagar can you confirm?

galipremsagar replied:
Yup, this will mark it exposed and copy-on-write buffers will be copied eagerly. This change looks good 👍

shwina commented on python/cudf/cudf/tests/test_df_protocol.py:
```diff
@@ -229,3 +268,12 @@ def test_NA_mixed_dtype():
+)
+def test_from_cpu_df(pandas_df):
+    df = pd.DataFrame({"a": [1, 2, 3], "b": ["x", "y", "z"]})
+    cudf.from_dataframe(df, allow_copy=True)
```
Thanks! I added some null tests.

> Also can we check the round-trip in the opposite direction? (Assuming that is supported too).

Today, that segfaults because Pandas doesn't check for the device type of the incoming buffers and tries to interpret the buffer pointers as host pointers :(

bdice commented on python/cudf/cudf/core/df_protocol.py:
```diff
@@ -645,28 +644,61 @@ def __dataframe__(
+
+    Parameters
+    ----------
+    df: Object supporting dataframe interchange protocol
+    allow_copy
+        If ``True``, allow copying of the data. If ``False``, a
+        ``TypeError`` is raised if data copying is required to
+        construct the ``DataFrame`` (e.g., if ``df`` lives in CPU
+        memory).
```
Fix the docstring syntax.
```suggestion
    df : DataFrameObject
        Object supporting dataframe interchange protocol
    allow_copy : bool
        If ``True``, allow copying of the data. If ``False``, a
        ``TypeError`` is raised if data copying is required to
        construct the ``DataFrame`` (e.g., if ``df`` lives in CPU
        memory).
```

bdice reviewed (approved):
One minor change for docstring syntax, otherwise LGTM.

wence- reviewed (approved):
Thanks, apologies for the delay

shwina commented:
/merge
